File: HM/python/HM_fit.py
# %%

def HM_fit(data,ro):
    from scipy.optimize import minimize
    import numpy as np
    from HM_lik import HM_lik

    xfit = []
    LL = []
    BIC = []
    nsubj= np.shape(data)[0]
    for ns in range(nsubj):
        def obFunc(params):
            x,y,z = params
            return HM_lik(data[ns,:], ro, x,y,z)
        x0 = np.array([0.5,4,0.5])
        bnds = ((0,1),(0,14),(0,1))
        res = minimize(obFunc, x0,method = 'Nelder-Mead',bounds=bnds)
        
        BIC.append(len(x0)*np.log(len(data[ns,:]))+2*res['fun'])
        LL.append(-res['fun'])
        xfit.append(res['x'])       
    xfit=np.array(xfit)
    return xfit,np.array(LL),np.array(BIC)

File: HM/matlab/HM_fit.m
function [Xfit, LL, BIC] = HM_fit(a, seq,ro)


matrix = size(a);
nsubj =matrix(1);

for ns = 1:nsubj
    obFunc = @(x) HM_lik(a(ns,:),seq, ro, x(1), x(2), x(3));
    
    X0 = [0.5,4,0.5];  % x 的初始点    
    % plot the objective function at each iteration
    options = optimset('PlotFcns',@optimplotfval);    
    [Xfit(ns,:), NegLL(ns,:)] = fminsearch(obFunc, X0, options);




    LL(ns,:) = -NegLL(ns,:);
    BIC(ns,:) = length(X0) * log(length(a)) + 2*NegLL(ns,:);
end

LL = LL
BIC = BIC
